benchmark/Naming.md

```diff
@@ -0,0 +1,148 @@
+
+````
+✅ Seq.dropFirst.Array
+✅ Seq.dropLast.Range.lazy
+✅ Seq.dropWhile.UnfoldSeq
+✅ Seq.prefix.AnySeq.RangeIter.lazy
+✅ Seq.prefixWhile.AnyCol.Array
+✅ Seq.suffix.AnySeq.UnfoldSeq.lazy
+
+✅ Existential.Array.ConditionalShift.Ref1
+✅ Existential.Array.Mutating.Ref2
+✅ Existential.Array.method.1x.Ref3
+✅ Existential.Array.method.2x.Ref4
+✅ Existential.Array.Shift.Val0
+✅ Existential.MutatingAndNonMutating.Val1
+✅ Existential.Mutating.Val2
+✅ Existential.method.1x.Val3
+✅ Existential.method.2x.Val4
+✅ Existential.Pass2.method.1x.Ref1
+✅ Existential.Pass2.method.2x.Ref2
+
+✅ Set.isSubset.Int25
+✅ Set.symmetricDifference.Int50
+````
+
+</details><p><!-- spacer --></p></li>
+<li>
+<strong>Groups and types are <code>UpperCase</code>, methods are
+<code>lowerCase</code>.</strong>
+<details>
+
+Use periods to separate the name components in variants derived from specialised
+generic types or significant method chains.
+
+````
+⛔️ InsertCharacterStartIndex
+⛔️ InsertCharacterTowardsEndIndexNonASCII
+````
+
+There's no need to be literal with type names. **Be descriptive**:
+
+````
+✅ Flatten.Array.Tuple4.lazy.flatMap
+✅ String.insert.ASCIIChar.StartIndex
+✅ String.insert.EmojiChar.NearEnd
+````
+
+</details><p><!-- spacer --></p></li>
+<li>
+<strong>Keep it short.</strong> 40 characters at most. Abbreviate if necessary!
+<details>
+
+Benchmarking results are reported on GitHub and very long names are causing
+horizontal table scrolling which unfortunately obscures the columns with actual
+measurements. Fixed upper size limit also helps with the formatted console
+output, when measuring locally. *It is more important for benchmark's name to be
+unique and short, than overly descriptive.*
+
+Prefer concise names for potential benchmark family extensions. Leave out the
+nested types from variants if they aren't strictly necessary for disambiguation.
+If there's potentially valuable future variant, like testing `ContinuousArray`,
+keep the `Array` now, allowing for addition of `ContArr` variants later.
+
+Use **`Val`** and **`Ref`** as short descriptors for variants that compare value
+types (`struct`, `Int`) with reference types (often named with `Class` in the
+legacy-style).
+Prefer **`Char`** to `Character`, which can be combined with codepage or
+language prefix/suffix when necessary (`ASCIIChar`). For benchmarks that measure
+`String`'s Unicode performance for various languages, use
+[two letter codes](https://en.wikipedia.org/wiki/ISO_639-1) instead of spelling
+out the whole language names.
+
+In a pinch, even C-style naming with two letter prefixes `OC` (for Objective-C)
+or abbreviations like  `Str` and `Arr` are OK, if it helps to fit a system with
+descriptive names into 40 characters:
```

Comment:
I don't work with the benchmarks much, but this doesn't match either LLVM or Swift naming conventions. I'd rather support having the benchmark runner insert ellipses into long names than telling people to start dropping vowels or syllables.

Reply:
This about coding convention, it's about readability of the report where we have a very long list of repetitive names. Abbreviations work well here.

Reply:
But we don't usually read the long list of names; only the ones that change come up.

Reply:
I agree with Jordan. I think it's okay to use abbreviations that are relatively widespread in the codebase anyway (e.g., ObjC for Objective-C), but inventing new ones hurts readability a lot:

Current: 

> ObjectiveCBridgeFromNSArrayAnyObjectForced

Better (although still not great):

> ObjectiveC.Bridge.From.NSArray.AnyObject.Forced 

Just as good: 

> ObjC.Bridge.From.NSArray.AnyObject.Forced

Worse:

> OC.Bridge.From.NSArr.AO.Forced

I hope we don't currently have anything as egregious as that last one.

Reply:
Fine with me, as long as parts of name can be easily elided in text-based tabular output.

Reply:
The `ObjectiveCBridge*` family of benchmarks is a good illustration of how to correctly apply the new principles. It has made me reconsider the legal set of characters to include `?` and `!`. The abbreviations I had in mind in that case @lorentey mentions would just mean replacing `ObjectiveCBridge` as a family prefix with `OCB`:
````
⛔️ ObjectiveCBridgeFromNSArrayAnyObjectForced
✅ OCB.NSArray.AnyObject.as!.Array.NSString
````

The details in the first bullet include it other members from that family in the example:
````
✅ OCB.NSArray.AnyObject.as?.Array.NSString
✅ OCB.NSArray.AnyObject.as!.Array.String
````
The point of **Keep it short** is to not be afraid to abbreviate where necessary, in order to fit the family into reasonable size, but to remain consistent. It's also illustrated in the details of the *Groups and variants* bullet.
````
✅ Seq.dropFirst.Array
✅ Seq.dropLast.Range.lazy
✅ Seq.dropWhile.UnfoldSeq
✅ Seq.prefix.AnySeq.RangeIter.lazy
✅ Seq.prefixWhile.AnyCol.Array
✅ Seq.suffix.AnySeq.UnfoldSeq.lazy
````
The local abbreviation rules in that benchmark family look like this:

| Type | Abbreviation |
|---|---|
| `AnyCollection` | `AnyCol` |
| `AnySequence` | `AnySeq` |
| `UnfoldSequence` | `UnfoldSeq` |
| `IndexingIterator<(Range<Int>)>` | `RangeIter` |

Should I rephrase the last point? Add more examples? I thought the document-as-is already contains it all, but apparently it isn't clear enough...

Reply:
The ideas about groups, name components and consistent abbreviations are lessons learned from introducing [extended coverage for sequence slicing operations](https://github.com/apple/swift/pull/8641). I tried my best with sequence names at the time, but when you pull all the results for comparison, [like I did here](https://forums.swift.org/t/troubling-sequence/13130) (look for table in under section **Cost of Abstractions, Revisited**), the newly proposed system can be used to slide and dice such a family much easier. When pulled together with new naming convention, the `Seq` family has 3 degrees of freedom: slicing **operation**, underlying sequence **type** and the eager/**lazy** dichotomy.

Reply:
The problem with using ad-hoc abbreviations like `OCB` is that they don't mean anything at all.

* "ObjectiveCBridgeFromNSArrayAnyObjectForced" is quite clumsy and not very readable, but if it pops up in a list of benchmark regressions I can look at the name and have a rough picture of what it measures.

* "OCB.NSArray.AnyObject.as!.Array.NSString" reads like an opaque mix of random characters and type names to me. I can see it has something to do with arrays and strings, but that's about it. I have to look up the source to see what it does.

The sequence slicing names are interesting; the abbreviations and camel casing do work pretty well together there. It hlps that the abbrevs are rltvly understable. But writing out the actual types in informal Swift syntax would still be a huge improvement.

I'm increasingly convinced these rules just get in the way of clarity. The goal is to make it clear what the benchmark measures -- so why not just write that down in whatever notation feels best?

```
ObjectiveCBridgeFromNSArrayAnyObjectForced
      → Bridging: NSArray<AnyObject> as! Array<NSString>

ObjectiveCBridgeStubFromArrayOfNSString2
      → Bridging: Stub from NSArray<NSString> (v2)

DropLastAnySeqCntRange
      → DropLast: AnySequence(CountableRange<Int>)
   or → AnySequence(CountableRange<Int>).dropLast()
    
DropFirstAnySeqCRangeIterLazy
      → DropFirst: AnySequence(CountableRange<Int>.Iterator).lazy
   or → AnySequence(CountableRange<Int>.Iterator).lazy.dropFirst()

RomanNumbers
      → Algorithm: Roman numeral conversion to/from Int
```

Reply:
I'm not sure we are trying to solve the same problem… The size of the benchmark names, when it comes to big families like `Existential` or `ObjectiveCBridge` are simply too long, and are making the reports unusable because of the horizontal scrolling. Names are used to invoke the individual benchmarks from the benchmark driver. I don't see how what you proposed(?) above addresses any of these issues.

Reply:
We definitely aren't! :) 

*My* pet problem is that benchmarks aren't labeled clearly, and I don't want them to get even worse. This is why I pushed back on enforcing camel casing, and this is why I dislike a-h abs & SATN. After all, TISNAEWTC!

On the other hand, having to scroll around on a webpage, or having to sometimes quote things on the command line honestly seem relatively minor issues to me. It is possible to design around those, while bad names remain bad no matter how well we display them.

Reply:
I'm working on integrating the [benchmark validation](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L264-L489) into the benchmark [command for CI](https://github.com/apple/swift/pull/20074). It checks that that **newly added** benchmarks conform to the best practices. For deviations it offers suggestions how to correct the issues.

Historically, one of the issues hampering usability of Swift Benchmark Suite were overlong test names, that were causing:
* misaligned columns in console output when running benchmarks with `build_script` and
* hiding the most important result columns in benchmark report tables on GitHub.

I have been trying to fix this problem for [more than year and a half](https://github.com/apple/swift/pull/9999#issuecomment-305296057). Back than, @atrick agreed that 40 characters would be a [reasonable limit:](https://github.com/apple/swift/pull/9999#issuecomment-305298530)
> The long benchmark names ruin my ability to table-ify the output, independent of Markdown. If benchmarks need verbose descriptions, we should add those. 40 chars is already ridiculous for a benchmark name.

Which is why  `BenchmarkDoctor`'s `check` has now a rule for [name length](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L324-L331). In order to maintain consistency among benchmarks names, I've also added rule for [naming convention](https://github.com/apple/swift/blob/a7f832fb57b459dc07c577db551bf88b39fa0629/benchmark/scripts/Benchmark_Driver#L311-L321), pointing to the `UpperCamelCase` in Swift Naming Guidelines.

@lorentey, you fairly [pushed against that](https://github.com/apple/swift/pull/20067#issuecomment-434663394) and [said (emphasis mine)](https://github.com/apple/swift/pull/20067#issuecomment-435482953):
> To me, **any separator** would be **much preferable** to `DictionaryOfAnyHashableStringsInsert`. `DictionaryOfAnyHashableStrings` is already quite bad, but appending Insert to it makes it unreadable.
> 
> Benchmark names are very similar to test names -- some of them are **hierarchical** we and **should have a rule** that accommodates that. **The dot would work very well!**
> 
> Names like `SetSymmetricDifferenceInt50` didn't bother me as much because their components are more straightforward, but `Set<Int>.symmetricDifference() 50%` would still read a lot better in any table. However, `Set.Int.symmetricDifference.50` is a **reasonable compromise** between the two, and it's still quite usable.

Incorporating your feedback I've proposed this naming convention, so that the `BenchmarkDoctor` can point to a concrete set of rules.

I hear that your main issue with benchmarks is clarity on what each of them does. Which is also fair, but I don't think it's realistic expectation to fully understand a benchmark just from its name. It's often the case that even looking at the source code leaves you without full comprehension of the original author's intent. Which is why @atrick's been pushing me to be [more diligent about documentation](https://github.com/apple/swift/pull/20067#issuecomment-435482953):

> For future benchmark maintenance, it's important to understand the justification and intention behind the benchmarks. Please add comments to test cases explaining what is being tested and why we care about performance of this particular test vs. more conventional ways to write it. I know nobody has really done this in the past, but I think it's important, especially when adding such a large number of strangely written tests.

Back to the proposed naming convention: the objections raised seem quite knee-jerk to me. It would take exactly one look at ObjectiveCBridging.swift to understand what `OCB` stands for. Repeating `ObjectiveCBridging` in the name of each test just clutters everything.

Allowing `?` and `!` in names is shortest notation describing how tests vary across two degrees of freedom — types being cast and optional/forced downcast. Avoiding the `Forced` suffix, allows adding **more information** about what and how is being tested to the name. Finishing the hypothetical renaming of tests from ObjectiveCBridging.swift according to proposed naming convention, I've realized that inclusion of `AnyObject` in the names was not adding information and caused name length overflow. Abbreviating `NSDictionary` to `NSDict` and `NSNumber` to `NSNum` made `OCB.NSDict.as!.Dictionary.NSString.NSNum` fit exactly the 40 characters limit. Here's the final result:

| After | Before |
|---|---|
| `OCB.NSString.as?.String` | `ObjectiveCBridgeFromNSString` |
| `OCB.NSString.as!.String` | `ObjectiveCBridgeFromNSStringForced` |
| `OCB.String.as.NSString` | `ObjectiveCBridgeToNSString` |
| `OCB.NSArray.as?.Array.NSString` | `ObjectiveCBridgeFromNSArrayAnyObject` |
| `OCB.NSArray.as!.Array.NSString` | `ObjectiveCBridgeFromNSArrayAnyObjectForced` |
| `OCB.Array.String.as.NSArray` | `ObjectiveCBridgeToNSArray` |
| `OCB.NSArray.as?.Array.String` | `ObjectiveCBridgeFromNSArrayAnyObjectToString` |
| `OCB.NSArray.as!.Array.String` | `ObjectiveCBridgeFromNSArrayAnyObjectToStringForced` |
| `OCB.NSDict.as?.Dictionary.NSString.NSNum` | `ObjectiveCBridgeFromNSDictionaryAnyObject` |
| `OCB.NSDict.as!.Dictionary.NSString.NSNum` | `ObjectiveCBridgeFromNSDictionaryAnyObjectForced` |
| `OCB.Dictionary.String.Int.as.NSDict` | `ObjectiveCBridgeToNSDictionary` |
| `OCB.NSDict.as?.Dictionary.String.Int` | `ObjectiveCBridgeFromNSDictionaryAnyObjectToString` |
| `OCB.NSDict.as!.Dictionary.String.Int` | `ObjectiveCBridgeFromNSDictionaryAnyObjectToStringForced` |
| `OCB.NSSet.as?.Set.NSString` | `ObjectiveCBridgeFromNSSetAnyObject` |
| `OCB.NSSet.as!.Set.NSString` | `ObjectiveCBridgeFromNSSetAnyObjectForced` |
| `OCB.Set.String.as.NSSet` | `ObjectiveCBridgeToNSSet` |
| `OCB.NSSet.as?.Set.String` | `ObjectiveCBridgeFromNSSetAnyObjectToString` |
| `OCB.NSSet.as!.Set.String` | `ObjectiveCBridgeFromNSSetAnyObjectToStringForced` |

Overall, I don't find the suggestion to allow free form names with spaces, brackets and braces remotely practical. Even allowing for `.` and `-` in the names [requires some modifications to the scripts that run benchmarks](https://github.com/apple/swift/pull/20667)… I admit that allowing `!` in the names would cross a line and require escaping (`\!`) or quoting the name when running such test individually in the command line. But it's still more practical and much more constrained design than freeform strings!

The whole workaround to overlong names of not showing all benchmark results in the CI reports on GitHub is IMHO a major regression in the transparency of Swift project. It's nice that you have Apple-internal performance tracking available, but since the the introduction of `run_smoke_bench`, there is no way for outside contributors to monitor performance of the Swift project without running the benchmarks locally. Previously I could search GitHub for the benchmark reports and get a sense of what's going on. I find smoke bench fine for what it does, but I think the ability to get full performance reports from CI should be re-established. Fixing the naming is therefore still important. Enforcing best practices for newly added tests is a first step.

Let's not throw out the baby with the bathwater here… What modifications to the proposed naming convention would make all of you agree that adding this to the project is beneficial?

Reply:
> I have been trying to fix this problem for more than year and a half. Back than, @atrick agreed that 40 characters would be a reasonable limit:

I also said that we can make do as long as there's some effective way to elide parts of the name. Once you have hierarchically organized tests and name separators, that limit would only apply to individual components of the name. I regularly parse and reformat long lists of tests in tabular text format. As long as there's a way to format the text output such that it's actually readable and can be automatically converted to a table, that works for me. 

e.g. output like this work work org mode's table-convert-region.
```
NAME                VAL, VAL, VAL
Bench1Group.
Bench1Name     #,  #,  #
Bench2Group.
Bench2Name     #, #, #
```

Reply:
I'll rephrase the explanation to stress the last resort nature of using abbreviations:

> *In a pinch*, even C-style naming with two letter prefixes `OCB` (for
Objective-C Bridging) or consistent abbreviations in the benchmark family like
`Str` and `Arr` are OK, but *only if* it helps to fit a system with descriptive
names into 40 characters:
> 
> ````
> ✅ CharCount.ContArr.Str.kr
> ✅ OCB.NSDict.as!.Dictionary.NSString.NSNum
> ✅ Seq.prefixWhile.AnySeq.UnfoldSeq.lazy
> ````

Reply:
Not restricting the length of a name is generally problematic, as it complicates effective design of other tools. Here’s the visualization of the whole benchmark suite runtimes I have been using to asses the health of the SBS: 

![benchmarks.html](https://user-images.githubusercontent.com/1046899/48694772-95e65480-ebdd-11e8-8f49-deb59edc810c.jpg)

In contrast to my [chart.html visualizations](https://palimondo.github.io/robust-microbench/chart.html?b=Calculator&v=iters), which work well on any mobile device thanks to responsive design, the overlong names are making benchmarks.html hard to use on smaller screens.